app/pipes.go

```diff
@@ -0,0 +1,174 @@
+package main
+
+import (
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/gorilla/mux"
+
+	"github.com/weaveworks/scope/xfer"
+)
+
+const (
+	gcInterval  = 30 * time.Second // we check all the pipes every 30s
+	pipeTimeout = 1 * time.Minute  // pipes are closed when a client hasn't been connected for 1 minute
+	gcTimeout   = 2 * time.Minute  // after 2 minutes, tombstoned pipes are forgotten
+)
+
+type pipeRouter struct {
+	sync.Mutex
+	wait    sync.WaitGroup
+	quit    chan struct{}
+	pipes   map[string]xfer.Pipe
+	lastPut map[string]time.Time // for timeing out pipes
+}
+
+func registerPipeRoutes(router *mux.Router) *pipeRouter {
+	pipeRouter := &pipeRouter{
+		quit:    make(chan struct{}),
+		pipes:   map[string]xfer.Pipe{},
+		lastPut: map[string]time.Time{},
+	}
+	pipeRouter.wait.Add(1)
+	go pipeRouter.gcLoop()
+	router.Methods("GET").Path("/api/pipe/{pipeID}").HandlerFunc(pipeRouter.handleUIPipeWS)
+	router.Methods("DELETE").Path("/api/pipe/{pipeID}").HandlerFunc(pipeRouter.deletePipe)
+	router.Methods("GET").Path("/api/pipe/{pipeID}/probe").HandlerFunc(pipeRouter.handleProbePipeWS)
+	return pipeRouter
+}
+
+func (pr *pipeRouter) stop() {
+	close(pr.quit)
+	pr.wait.Wait()
+}
+
+func (pr *pipeRouter) getOrCreatePipe(id string, track bool) (xfer.Pipe, bool) {
+	pr.Lock()
+	defer pr.Unlock()
+	pipe, ok := pr.pipes[id]
+	if !ok {
+		log.Printf("Creating pipe id %s", id)
+		pipe = xfer.NewPipe(id)
+		pr.pipes[id] = pipe
+	}
+	if pipe.Closed() {
+		return nil, false
+	}
+	if track {
+		delete(pr.lastPut, id)
+	}
+	return pipe, true
+}
+
+func (pr *pipeRouter) gcLoop() {
+	defer pr.wait.Done()
+	ticker := time.Tick(gcInterval)
+	for {
+		select {
+		case <-pr.quit:
+			return
+		case <-ticker:
+		}
+
+		pr.timeoutPipes()
+		pr.gcPipes()
+	}
+}
+
+func (pr *pipeRouter) timeoutPipes() {
+	pr.Lock()
+	defer pr.Unlock()
+	now := time.Now()
+	for id, lastPut := range pr.lastPut {
+		if now.Sub(lastPut) > pipeTimeout {
+			log.Printf("Timeing out pipe %s", id)
+			delete(pr.lastPut, id)
+			pipe := pr.pipes[id]
+			pipe.Close(now)
+		}
+	}
+}
+
+func (pr *pipeRouter) gcPipes() {
+	pr.Lock()
+	defer pr.Unlock()
+	now := time.Now()
+	for pipeID, pipe := range pr.pipes {
+		if pipe.Closed() && now.Sub(pipe.ClosedAt()) > gcTimeout {
+			log.Printf("GCing pipe %s", pipeID)
+			delete(pr.pipes, pipeID)
+		}
+	}
+}
+
+func (pr *pipeRouter) handleUIPipeWS(w http.ResponseWriter, r *http.Request) {
+	var (
+		vars   = mux.Vars(r)
+		pipeID = vars["pipeID"]
+		method = r.FormValue("_method")
+	)
+	// UI uses a sendBeacon to ensure pipes are closed when the UI exits,
+	// but sendBeacon can only POST; this work around is inspired by rails.
+	if method == "delete" {
+		pr.deletePipe(w, r)
+		return
+	}
+
+	pipe, ok := pr.getOrCreatePipe(pipeID, true)
+	if !ok {
+		log.Printf("Pipe %s is not connected right now...", pipeID)
+		http.NotFound(w, r)
+		return
+	}
+	defer func() {
+		pr.Lock()
+		defer pr.Unlock()
+		pr.lastPut[pipeID] = time.Now()
```

Comment:
My understanding of this, is it will timeout any pipes 10 minutes after a UI disconnects (so long as there has not been another connection established since then). Correct?

Is there a case like this?
- ui1 connects
- ui2 connects
- ui1 disconnects
- pipe is gced and ui2 is booted